katsdpsigproc/tune.py

```diff
@@ -139,49 +146,31 @@ def _query(
         Keys and values for the query
     """
     try:
-        query, query_args = _assemble_query(tablename, keys)
-        cursor = conn.cursor()
-        cursor.execute(query, query_args)
-        row = cursor.fetchone()
-        # if a match is not found, find the nearest match by first ignoring the device version
-        if row is None:
-            _logger.debug(
-                "Retrying query '%s' by ignoring device_version", query, exc_info=True
-            )
-            try:
-                del keys["device_version"]
-            except KeyError:
-                pass
-            query, query_args = _assemble_query(tablename, keys)
-            cursor = conn.cursor()
-            cursor.execute(query, query_args)
-            row = cursor.fetchone()
-        # if a match is still not found, next try ignoring the device platform
+        query, query_args, row = _assemble_query(conn, tablename, keys)
         if row is None:
-            _logger.debug(
-                "Retrying query '%s' by ignoring device_platform", query, exc_info=True
-            )
-            try:
-                del keys["device_platform"]
-            except KeyError:
-                pass
-            query, query_args = _assemble_query(tablename, keys)
-            cursor = conn.cursor()
-            cursor.execute(query, query_args)
-            row = cursor.fetchone()
-        # if a match is still not found, next try ignoring the device name
-        if row is None:
-            _logger.debug(
-                "Retrying query '%s' by ignoring device_name", query, exc_info=True
-            )
-            try:
-                del keys["device_name"]
-            except KeyError:
-                pass
-            query, query_args = _assemble_query(tablename, keys)
-            cursor = conn.cursor()
-            cursor.execute(query, query_args)
-            row = cursor.fetchone()
+            if KATSDPSIGPROC_TUNE_FALLBACK == "fail":
+                raise RuntimeError(
+                    "Did not find record and autotune fallback behaviour set to 'fail'."
+                )
```

Comment:
Thinking about this some more I reckon my initial proposal is a bit flawed: it conflates matching behaviour (how hard to search to find an existing set of parameters) with fallback behaviour (what to do if there is no match). Sorry to move the goalposts, but I'd suggest changing the variable to be called `KATSDPSIGPROC_TUNE_MATCH`, with options `exact` and `nearest`. I'd say for now don't worry about a knob for the behaviour when no match is found.